### DISCUSSION TITLE
How  to handle Maven CI Friendly Versions?

### DISCUSSION THREAD

**Discussion by kjq** [renovatebot/renovate#19423] [Request Help] [answered]
I have a question about using Renovate with [Maven CI friendly versions](https://maven.apache.org/maven-ci-friendly.html).

Renovate recognizes the `-SNAPSHOT` or release properly but instead of replacing the property with the full version (i.e. `1.2.3-SNAPSHOT`) we would instead want to replace the property with only `1.2.3` because an additional property is used or blanked-out for the release.

From the Maven side, this looks like:
```xml
<properties>
    <revision>1.2.3</revision>
    <changelist>-SNAPSHOT</changelist>
</properties>
```

```xml
<parent>
    <groupId>some-group</groupId>
    <artifactId>som-parent</artifactId>
    <version>${revision}${changelist}</version>
</parent>
```

Regardless of whether it is a SNAPSHOT or release all that we update is the `<revision>` because the rest is handled from the CI side.

Is there a way to update the `<revision>` property with the version# only for this scenario even though the dependency is recognized as the full version?

**Comment by rarkins** [comment]
Sorry, could you give more details to help my understanding? If possible:

- What it looks like before
- What it looks like today when Renovate upgrades it
- What you would prefer it looks like

**Comment by kjq** [comment]
Sure, the POM will use properties:

```
<properties>
   <revision>5.3.0</revision>
   <changelist>-SNAPSHOT</changelist>
</properties>
```

The effective POM will see this as `5.3.0-SNAPSHOT`  

When a new version is published, `-SNAPSHOT` or a release it _only_ updates the `revision` part with that version (i.e. `5.3.1`) even though the release is formally `5.3.1-SNAPSHOT` coming from the repository.

```
<properties>
   <revision>5.3.1</revision>
   <changelist>-SNAPSHOT</changelist>
</properties>
```

Builds can be performed "as snapshots" with `mvn install` or releases with `mvn install -Dchangelist=` (remove the suffix).

It is used within the POM like this `<version>${revision}${changelist}</version>`.

So, fo example, another POM where that is the parent would look like:

```
<parent>
     <groupId>parent-group</groupId>
    <artifactId>parent-pom</artifactId>
    <version>${revision}${changelist}</version>
</parent>
```

Essentially, instead of changing the version in POM's to `5.3.1-SNAPSHOT` I would only change that `revision` to the version# alone.

**Reply by rarkins** [reply]
Do you mean that Renovate does not update this at all today, and you'd like it to update them?

**Comment by kjq** [comment]
No, Renovate does what it should as if this is a conventional Maven POM, it updates the `<version>` as `5.3.1-SNAPSHOT`.  I need to alter this update to only change the `revision` property with just the version instead of the whole thing.

**Reply by rarkins** [reply]
What does the `packageFiles with updates` log message look like for this dependency?

**Reply by kjq** [reply]
I will have to make a run with Renovate and logging and will post it today or tomorrow...

**Comment by kjq** [comment, marked as answer]
Closing because we were able to do this by relying on Maven as-is.

Since the property is _just_ the version and we recently moved to more of a release strategy rather than using snapshots, but the CI system builds both the release and the snapshot, Renovate works properly and updates the version as it should using the release (`X.y.z`).

In Maven, the `<changelist>-SNAPSHOT</changelist>` does not change, only the `<revision>X.y.z</revision>`
